### PR TITLE
Add support for hash fragment redirect targets

### LINKED ISSUE #16: Add support for hash fragments
__First of all: this is a very useful plugin, thanks for writing it!__

I'm currently in the process of rewriting the docs for Material for MkDocs in https://github.com/squidfunk/mkdocs-material/pull/1735 and, as the overall structure changed a lot, want to redirect some of the pages. However, some pages were merged.

## Idea

Allow hash fragments in URL mappings, so that the redirect will jump to the subsection of the page, e.g.:

``` yaml
plugins:
  - search
  - redirects:
      redirect_maps:
        releases/4.md: upgrading.md#upgrading-to-4-x
        releases/5.md: upgrading.md#upgrading-to-5-x
        releases/changelog.md: changelog.md
```

## Actual behavior

When building the site, the plugin prints:

```
WARNING -  Redirect target 'upgrading.md#upgrading-to-4-x' does not exist! 
WARNING -  Redirect target 'upgrading.md#upgrading-to-5-x' does not exist! 
```

## Desired behavior

No error, redirects work.

### PR DESCRIPTION
**Pull request by wwarriner**
Adds support for hash fragment redirect targets.

This will allow e.g. `old.md: new.md#section` to work as expected.

I added a helper function to split the hash fragment off the rest of the path. Whenever the hash fragment gets in the way of path manipulation logic, it gets detached and/or reattached, as appropriate.

To validate I added duplicate test cases with hash fragments, though there are no tests for `on_post_build()`. I verified that with a mkdocs project, but can't readily share that verification.

Fixes #16

### CODE REVIEW

File: mkdocs_redirects/plugin.py [hunk 1]
```diff
@@ -125,12 +130,13 @@ def on_post_build(self, config, **kwargs):
 
         # Walk through the redirect map and write their HTML files
         for page_old, page_new in self.redirects.items():
+            new_path_hash_free, _ = _split_hash_fragment(str(page_new))
```
Comment (oprypin) [inline mkdocs_redirects/plugin.py]: Please let's not silently drop any part. This is just confusing. Kinda unrelated change, anyway.

File: mkdocs_redirects/plugin.py [hunk 2]
```diff
@@ -65,31 +66,35 @@ def get_relative_html_path(old_page, new_page, use_directory_urls):
...
         else:
-            return posixpath.join(parent, name, 'index.html')
+            html_path_segments = [name, index_name]
 
     # Just use the original name if Directory URLs aren't used
     else:
-        return posixpath.join(parent, (name + '.html'))
+        html_path_segments = [name + '.html' + hash_fragment]
+
+    return posixpath.join(parent, *html_path_segments)
```
Comment (oprypin) [inline mkdocs_redirects/plugin.py]: This can be done much simpler. Assuming all old code above, just replace `return` with `new_path = `.
And then
```suggestion
    return path + hash_fragment
```

File: mkdocs_redirects/plugin.py [hunk 3]
```diff
@@ -143,3 +149,20 @@ def on_post_build(self, config, **kwargs):
...
+    "path/to/file"      => ("/path/to/file", "")
+    """
+    if '#' in path:
+        parts = path.split('#', maxsplit=1)
+        path = parts[0]
+        hash_fragment = '#' + parts[1]
+    else:
+        hash_fragment = ''
+
+    return path, hash_fragment
```
Comment (oprypin) [inline mkdocs_redirects/plugin.py]:
```suggestion
    path, hash, after = path.partition('#')
    return path, hash + after
```

File: mkdocs_redirects/plugin.py [hunk 4]
```diff
@@ -54,7 +54,8 @@ def write_html(site_dir, old_path, new_path):
 def get_relative_html_path(old_page, new_page, use_directory_urls):
     """ Return the relative path from the old html path to the new html path"""
     old_path = get_html_path(old_page, use_directory_urls)
-    new_path = get_html_path(new_page, use_directory_urls)
+    new_path_with_hash = get_html_path(new_page, use_directory_urls)
+    new_path, new_hash_fragment = _split_hash_fragment(new_path_with_hash)
```
Comment (oprypin) [inline mkdocs_redirects/plugin.py]: Why not just put these two lines in the opposite order and then `get_html_path` doesn't need to also have the splitting logic

File: mkdocs_redirects/plugin.py (hunk 1 above)
Comment (wwarriner) [reply mkdocs_redirects/plugin.py]: All the other suggestions make perfect sense to me and work great, and I've incorporated them locally. Thank you.

Without this change (and the associated change on line 139), I get a warning about non-existent target pages when building documentation using `mkdocs serve` and an HTTP 404 error when navigating to the source URL in the browser. The warnings and errors are exclusive to targets with hash fragments.

With this change there is no warning and the source URL correctly redirects to the target.

I'm open to suggestions on improving clarity, or a better place in the code to make this change.

Comment (oprypin) [reply mkdocs_redirects/plugin.py]: Ah yes, this suggestion doesn't make sense. I thought it's changing page_old but it's changing page_new - that's all good.

**Comment by wwarriner** [comment]
I think this is done. I am unsure how to address the concern over silent drop of the second return value from the helper function on line 133 on `plugin.py`. In my experience this hasn't caused any issues, but I defer to your expertise.

**Review by oprypin — APPROVED** [review]

**Comment by oprypin** [comment]
Thanks

**Comment by oprypin** [comment]
:heavy_check_mark: :shipit:   https://github.com/mkdocs/mkdocs-redirects/releases/tag/1.1.0
